### ISSUE TITLE
feature request: raysession should restore window sizes and positions (geometry) on session restore

### ISSUE THREAD

**Issue by N0rbert** [Houston4444/RaySession#105]
Thank you for your great project!

It would be really great if `raysession` will save and restore window sizes and positions (geometry).
If I understand correctly, then the current master branch only saves the geometry, but can't restore it.

**Comment by Houston4444** [comment]
Hi. Thanks you.
What are you talking about ? The RaySession window itself ? or the clients (programs) windows managed by RaySession ?
If it concerns RaySession window itself, it saves and restores the window geometry, if it doesn't works for you, then this is a bug. Which desktop environnement do you use ? Are you sure that RaySession is the only one Qt application where it does not works correctly ?

**Comment by N0rbert** [comment]
Thank you for reply. I'm very new to RaySession. If I'm doing something wrong, please correct me.

I'm talking about the clients (programs) windows managed by RaySession. Below is the screencast of what I mean:

![out](https://user-images.githubusercontent.com/309456/108811521-7b6e6300-75be-11eb-9ea4-7cb7831edd84.gif)

I expect that window positions will be restored on session start.

I'm using Ubuntu MATE 20.04.1 LTS with MATE 1.24.0, raysession is from deb-package at version 0.8.3-0ubuntu2. The above happens with and without Desktops Memory option enabled.

**Comment by N0rbert** [comment]
After some trial and error I have also tried to install latest RaySession version from master branch. To keep this process reproducible I have [documented it on AskUbuntu](https://askubuntu.com/q/1318585/66509).

Of course I have installed `wmctrl` executable to the system. I have all options enabled - *Desktops Memory*, *Session scripts*, *Remember optional GUI states*.

What am I doing wrong? Is it a bug?

**Comment by Houston4444** [comment]
Ok. So this is not a bug, this is something not developed.
The only one daemon option referring to what you are talking about is "Desktops Memory", all the others ones have nothing to do with that. Read [the manual](http://raysession.tuxfamily.org/en/manual.html#_the_daemons_options) if you want to know what they do exactly.

**Desktops Memory** only remembers the virtual desktop where were client windows. The reason why it doesn't tries to restore windows positions IIRC is that it is really too much work to deal with different window managers not reacting the same. More, linux desktops are switching (slowly) from X server to Wayland where dealing with windows programmaticaly becomes harder.

IIRC, there was also a problem with wmctrl not maximizing windows correctly, and probably a problem with multiple screens.

**Comment by M0JXD** [comment]
After a lengthy discussion in https://github.com/M0JXD/non-sequencer-fork/issues/3 this is feature I'd like to see in RaySession too, although it's an understandably difficult feature to implement.

**Comment by Houston4444** [comment]
> After a lengthy discussion in https://github.com/M0JXD/non-sequencer-fork/issues/3 this is a feature I'd like to see in RaySession too, although it's an understandably difficult feature to implement.

Indeed, I think it is too much work to have it nicely working with all existing window managers. It means that I won't try to do it myself, I tried years ago with wmctrl, it was my intention when I develop "desktops memory" option, I don't think the situation is easier now (I may be wrong). 

Of course, if someone has the motivation to develop a such feature, it could be added. It's probably partially doable with a session script (in load.sh), with the help of wmctrl or maybe dbus messages for Wayland window managers.

**Comment by M0JXD** [comment]
In the mentioned issue there's discussion of a whether a specialised client for this akin to jackpatch (although RaySession has it's own version anyways) is a worthwhile solution. That way it could work for any choice of session manager.

But yes especially given we are partway the steady transition to Wayland it can be awkward.

Maybe using XSM and the specific Gnome/KDE ones under Wayland? https://en.wikipedia.org/wiki/X_session_manager#Implementations

**Comment by Houston4444** [comment]
> In the mentioned issue there's discussion of a whether a specialised client for this akin to jackpatch (although RaySession has it's own version anyways) is a worthwhile solution. That way it could work for any choice of session manager.

Mmmh, I think it can't works perfectly this way. This client should be aware of client status (and their pid probably), few clients are able to broadcast and RaySession can either load non NSM apps. For me, it is definitely a server side option.

> But yes especially given we are partway the steady transition to Wayland it can be awkward.

We can't be even sure that the transition will totally be done one day. When I started this project, transition was already started, but some users are frustrated with missing features by design (I understand them, xdotools and wmtcrl can do very nice things).
